deepeval/evaluate.py

```diff
@@ -69,31 +71,96 @@ def create_api_test_case(
+        cache_key = json.dumps(sorted_cache_data)
+        
+        return cache_key 
+
 def execute_test_cases(
     test_cases: List[LLMTestCase],
     metrics: List[BaseMetric],
     save_to_disk: bool = False,
 ) -> List[TestResult]:
+    
     test_results: List[TestResult] = []
     test_run_manager.save_to_disk = save_to_disk
+
     for index, test_case in enumerate(test_cases):
+
+        test_run = test_run_manager.get_test_run()
+        cached_test_run = test_run_cache_manager.get_cached_test_run()
+        key = generate_cache_key(test_case=test_case)
+        look_up_map = cached_test_run.test_cases_lookup_map
+
+        #########################################
+        # Get cached data if available
+        ##########################################
+        already_in_cache = (key in look_up_map)
+        hyperparameter_same = cached_test_run.hyperparameters == test_run.hyperparameters
+        actual_output_exists = False
+        metrics_metadata_exists = False
+        cached_api_test_case = None
+
+        if already_in_cache:
+            cached_api_test_case = look_up_map[key]
+            actual_output_exists = cached_api_test_case.actual_output != None
+            if actual_output_exists:
+                metrics_metadata_exists = cached_api_test_case.metrics_metadata != None
+        can_use_cache = (already_in_cache and hyperparameter_same and actual_output_exists and metrics_metadata_exists)
+
+        cached_metrics_metadata = {} 
+        if can_use_cache:
+            metrics_metadata = look_up_map[key].metrics_metadata
+            cached_metrics_metadata = {metadata.metric: metadata for metadata in metrics_metadata}
+
+        #########################################
+        # Calculate remaining metrics (if not cached)
+        #########################################
         success = True
         api_test_case: APITestCase = create_api_test_case(test_case, index)
         test_start_time = time.perf_counter()
 
+
         for metric in metrics:
-            # Override metric async
-            metric.async_mode = False
-
-            metric.measure(test_case)
-            metric_metadata = MetricsMetadata(
-                metric=metric.__name__,
-                score=metric.score,
-                threshold=metric.threshold,
-                reason=metric.reason,
-                success=metric.is_successful(),
-                evaluationModel=metric.evaluation_model,
-            )
+            metric_metadata = None
+            if metric.__name__ in cached_metrics_metadata:
+                metric_metadata = cached_metrics_metadata[metric.__name__]
+            
+            # If same metric_metadata does not exist in cache 
+            # or threshold is not in same as in cached
+            # recomputer metric
+            new_threshold = metric.threshold
+            if not metric_metadata or new_threshold != metric_metadata.threshold:
```

Comment:
Can abstract this logic in a helper function, since you also have to check for things like include_reason, same evaluation model, and strict mode